### PR TITLE
codeintel: Optimize commit graph

### PR DESCRIPTION
**Pull request by efritz**
This is an effort to reduce memory pressure as described in https://github.com/sourcegraph/sourcegraph/issues/16030.

There are a few changes here:

1. Instead of maintaining sets of visible uploads with a structure shaped like `map[string][]UploadMeta`, we keep track of it in `map[string]map[string]UploadMeta`. This is possible as when looking in one direction down the commit graph there will only be at most one upload visible with a given indexer and root value. This greatly speeds up the cases in which a repository has many distinct roots, as we no longer have to do a linear search over each commit visible at a commit. This comes at a slight cost of memory, which we continue to optimize.

@Strum355 has suggested either making the remaining map into another map with smaller, more efficient keys (string -> int conversion), and also suggested a radix tree-like structure (or a variant such as a Patricia tree) in order to cut down on the overhead fo storing maps. These may be interesting ideas, but aren't urgently necessary due to the next change.

2. Instead of pre-calculating the set of visible uploads for _every_ commit at once, we only calculate the set of visible uploads for some class of commits. The benchmark data added in this PR comes from a customer that was experiencing scale problems. In this case, we only need to store data for 20% of the commit graph. The remaining commits can be calculated on the fly with a cheap traversal of the commit graph (more technical explanations in code comments), meaning that we only need to keep data for 20% of the graph resident in memory.

Benchmark results:

```
  Before: 17998614217551 ns/op   21031275248 B/op   1019199 allocs/op   17999.282s (5 hours)
Progress:    48014817436 ns/op   21558759232 B/op   1849823 allocs/op      49.893s
Progress:    23638566834 ns/op    9836824888 B/op    334060 allocs/op      24.604s
   After:     4960440863 ns/op    1139648984 B/op    122357 allocs/op       5.443s
```

### CODE REVIEW

File: enterprise/internal/codeintel/stores/dbstore/dumps.go [hunk 1]
```diff
@@ -225,10 +225,7 @@ func (s *Store) FindClosestDumpsFromGraphFragment(ctx context.Context, repositor
 		return nil, err
 	}
 
-	visibleUploads, err := calculateVisibleUploads(graph, commitGraphView)
-	if err != nil {
-		return nil, err
-	}
```
Comment (daxmc99) [inline enterprise/internal/codeintel/stores/dbstore/dumps.go:231]: line 285 of this file is unused (according to golangci-lint) 👀

Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/dumps.go:231]: Used in line 231?

Comment (daxmc99) [reply enterprise/internal/codeintel/stores/dbstore/dumps.go:231]: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/codeintel/stores/dbstore/dumps.go?subtree=true#L285:6&tab=references 
Looks like this function is unused, might be nice to clean it up here

Comment (daxmc99) [reply enterprise/internal/codeintel/stores/dbstore/dumps.go:231]: I can't comment on the line because it's not part of the diff :p

Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/dumps.go:231]: Oh sorry I totally missed the line number. Yeah this is left-over, I'll make a task for a dead code removal pass.

**Review by Strum355 — APPROVED** [review]

**Pull request merged**